src/Type.hs

```diff
@@ -25,18 +25,25 @@ newMetaPtr = MetaPtr <$> newUnique
 instance Show MetaPtr where
   show (MetaPtr i) = "(MetaPtr " ++ show (hashUnique i) ++ ")"
 
-data TyF t
+data TypeConstructor
```

Comment:
If we wanted to, we could index `TypeConstructor` by the number of type arguments, and get some more static safety for the built-in ones. I don't think it's worth it, but it is definitely a possibility.

And we could do something like this:
```
{-# LANGUAGE DataKinds #-}

data ArgCount = None | One | Two | Many
data TypeConstructor :: ArgCount -> Type where
  TSyntax :: ArgCount None
  -- ...
  TFun :: ArgCount Two
  -- ...
  TDatatype :: ArgCount Many

data TyApp =
  App0 (TypeConstructor None) | App1 (TypeConstructor One) (TyApp) | App2 (TypeConstructor Two) TyApp TyApp | AppMany (TypeConstructor Many) [TyApp]
```
This basically treats TypeConstructor as one type per number of args, with the advantage of being able to create a little existential package when we want to talk about them generically rather than having to build the explicit sum type (basically using the iso between Either A B and (Sigma b : Bool . if b then A else B).

This is probably not worth the added complication, but you had mentioned being a bit unhappy with the lack of static safety so I though it was worth mentioning.